### PR TITLE
integration: integration: ensure leader is up in waitLeader() and clusterMustProgress()

### LINKED ISSUE #7258: TestRestartMember: etcd member has no leader
via semaphore https://semaphoreci.com/coreos/etcd/branches/master/builds/1505

```
--- FAIL: TestRestartMember (6.42s)
	cluster_test.go:454: create on unix://127.0.0.1:2265625374 error: client: etcd member unix://127.0.0.1:2265625374 has no leader
```

### PR DESCRIPTION
**Pull request by fanminshi**
The issue is caused by leader loss even after waitLeader() returns
which can happen if the test machine is flaky which triggers a leader loss
or the killed node is the leader since waitLeader() only scans followers in
TestRestartMember() and they can have the same older leader.
In those cases, clusterMustProgress() proceeds with no leader which triggers
the no leader error.

To get around that, use linearizable get in waitLeader() to ensure leader is up
and retries on kapi.create() in clusterMustProgress() to ensure it proceeds with
a leader.

FIX #7258

### CODE REVIEW

**Comment by xiang90** [comment]
Can you explain why do we need the retry here? Are you able to reproduce the problem?

File: integration/cluster_test.go [hunk 1]
```diff
@@ -449,7 +449,17 @@ func clusterMustProgress(t *testing.T, membs []*member) {
...
+	var (
+		err  error
+		resp *client.Response
+	)
+	for i := 0; i < 3; i++ {
+		resp, err = kapi.Create(ctx, "/"+key, "bar")
+		if err == nil {
+			break
+		}
+		t.Logf("WARN: create on %s error: %v. retry again ...", membs[0].URL(), err)
```
Comment (heyitsanthony) [inline integration/cluster_test.go]: idiomatic go messages are all lower case:
`t.Logf("failed to create key on %q (%v)", membs[0].URL(), err)`

File: integration/cluster_test.go [hunk 2]
```diff
@@ -449,7 +449,17 @@ func clusterMustProgress(t *testing.T, membs []*member) {
 	kapi := client.NewKeysAPI(cc)
 	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
```
Comment (heyitsanthony) [inline integration/cluster_test.go]: this needs to be inside the for loop or the retry will end up using an expired context (also `cancel()` after making the `Create` call)

**Comment by fanminshi** [comment]
@xiang90 I produce the issue only if I commented out [`c.waitLeader(t, membs)`](https://github.com/coreos/etcd/blob/master/integration/member_test.go#L57).  I suspect that the issue is caused by leader loss even after `waitLeader()` returns which can happen if the test machine is flaky which trigger a leader loss or the killed node is the leader since `waitLeader()` only scan followers and they can have the same older leader. In those cases, `clusterMustProgress()` proceeds with no leader which triggers the `no leader`error.

To get around that, adding retry on `create` to ensure that leader is up before proceeding rest of the `clusterMustProgress()` code.

**Comment by fanminshi** [comment]
All fixed. PTAL @heyitsanthony

**Comment by xiang90** [comment]
@fanminshi Can you update the commit message with the explanation?

**Comment by fanminshi** [comment]
@xiang90 will do!

**Comment by xiang90** [comment]
> waitLeader() only scan followers and they can have the same older leader

If this is what the current logic does, I feel it worth fixing. We need to ensure to wait for a new leader coming up in a few cases.

**Comment by fanminshi** [comment]
@xiang90 sounds good. I'll fix waitLeader().

**Comment by fanminshi** [comment]
All fixed. PTAL @xiang90 @heyitsanthony

**Comment by fanminshi** [comment]
TestIssue3699 is not passing. Investigating.

File: integration/cluster.go [hunk 3]
```diff
@@ -343,6 +347,19 @@ func (c *cluster) waitLeader(t *testing.T, membs []*member) int {
 	for _, m := range membs {
 		possibleLead[uint64(m.s.ID())] = true
 	}
+	c.URLs()
```
Comment (fanminshi) [inline integration/cluster.go]: yeah, it was a mistake.

File: integration/cluster.go [hunk 4]
```diff
@@ -175,8 +175,12 @@ func (c *cluster) URL(i int) string {
 
 // URLs returns a list of all active client URLs in the cluster
 func (c *cluster) URLs() []string {
+	return GetMembersURLs(c.Members)
+}
+
+func GetMembersURLs(Members []*member) []string {
```
Comment (fanminshi) [inline integration/cluster.go]: sounds good. will change

**Comment by fanminshi** [comment]
All fixed. PTAL cc\ @heyitsanthony

**Comment by heyitsanthony** [comment]
lgtm. Thanks!

**Pull request merged**